wincent/liferay-portal
# LPS-114796 Update modal usages in layout modules

Hey @wincent 

In this PR we are updating modals in `layout` modules, extending `openModal` API where needed, and fixing few bugs on the way.

Notes:
- By far the biggest complication in this task was the update of [modal](https://github.com/liferay/liferay-portal/blob/60899f1e0d974932c2ed4409bdff03568a651197/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/processors/HTMLProcessor.js#L31) in `HTMLProcessor.js`. We had plenty of complex iframe modal usages, this is the first complex non-iframe one:
  - API needed to expanded with `button.onClick`
  - We needed to define a way to close modal on demand. It used to use `dialog.hide()`, but we no longer have component instance. In the PR, I am using the `closeModal` custom event to achieve this. I initially added a closing function as `onClick` argument,
```onClick: ({closeModal}) => {... closeModal()} ```,
but then it turned out we need the custom event in [a different place](https://github.com/wincent/liferay-portal/commit/7e531d4efe45dc879248eacbe02a37a6950cea90#diff-981f5eb37ee1df0062b69d69639ff672R163). Although I would consider passing capability over argument a better solution, this way we are more consistent, and we don't need to maintain two points of API that do the same thing.
  - It uses modal without `url` or `bodyHTML`. This caused `onOpen` not to work correctly. I started coding in support for this case, but it turned out it would require us to have a new component on [this line](https://github.com/liferay/liferay-portal/blob/60899f1e0d974932c2ed4409bdff03568a651197/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js#L267), causing a lot of refactoring, and another level of component depth. It would be `Modal` > `New Component` > `Iframe`. We can do this, but I opted to go with much simpler route and add an empty container in `bodyHTML`. For this modal, we need an extra container in any case.
  - Bigger padding of Clay modal caused CSS issues, namely because of the awkward 95% height on editor. The style changes in this PR override this.
- We are [removing](https://github.com/wincent/liferay-portal/commit/1525e0828da2c8f43a3e089455be961c5cae19b0#diff-45c988fca3398d3143e41f2f53e309a7R187) `modal-open` class on close. We had this bug from the start, but it only became noticeable in the case of returning to a page with content large enough to overflow the screen. We would no longer be able to scroll page.
- [If `id` was not set](https://github.com/wincent/liferay-portal/commit/1525e0828da2c8f43a3e089455be961c5cae19b0#diff-45c988fca3398d3143e41f2f53e309a7R303), we would render `id="undefined_iframe"` @jbalsas 

Test Plan:

1. Site Builder > Pages > (+) > Widget Page
- intentional change in appearance, from fixed width and height to full screen
---
1. Site Builder > Pages > Page item options > Permissions
---
1. Site Builder > Pages > Public Pages > cog icon
1. Change Current Theme
1. Confirm page is scrollable after theme selection
---
1. Site Builder > Pages > Page item options > Copy Page
- intentional change in appearance, from fixed width and height to full screen
---
1. Site Builder > Pages > Page item options > Edit
1. Add a widget to the page
1. Widget item > Permissions
- the appearance is a bit buggy, but that's unrelated to this PR. For some reason, we are including `_custom_properties_set.scss` and this messes up styles. Other permission modals look fine. I can open up a bug for this.
---
1. Make a sample Web Content
1. Site Builder > Pages > Page item options > Edit
1. Add a Web Content Display widget to the page
1. Set it to show sample the Web Content, refresh
1. Right sidebar > Contents > sample web content options > Permissions
---
1. The same as previous
1. Right sidebar > Contents > sample web content options > View Usages
---
1. Site Builder > Pages > Page item options > Edit
1. Add a HTML fragment to the page
1. Click twice on HTML example text > HTML fragment options > Edit
- this is the usage that caused a lot of headache 
---
1. Site Builder > Page Templates > (+) > Make a sample master page
1. Master page item options > Permissions
---
1. Site Builder > Page Templates > Master page item options > Export
1. Site Builder > Page Templates > Top right options  > Import
1. Import the exported file
---
1. Site Builder > Page Templates > Page Templates tab > (+) > Make a collection
1. Collection item > (+) > Make a Content Page Template
1. Content Page Template item options > Permissions
---
1. Site Builder > Page Templates > Display Page Templates tab > (+) > Make a sample template
1. Template item options > Permissions
---
1. Global site > Site Builder > Page Templates > item options > Permissions
---
1. Global site > Site Builder > Page Templates > item options > Import
---
1. Global site > Site Builder > Page Templates > item options > Export
---
1. Content & Data > Web Content > (+) > Make a web content
1. Home Page > Add a Web Content Display App
1. Configure Web Content Display App to show the web content
1. Content & Data > Web Content item > Edit
1. Right side menu > Usages tab > Eye icon
- If the product menu was open when you did this, it shows in preview as an empty block. The same behavior is in master. I can open a bug for this.
---
I was unable to find flattened_view.jsp > layout_action.jsp usage. It is the same 'copy page' option as previous. It may be that this is unused, as `copyLayoutDialog` does not appear anywhere else. In the previous usage, `addLayoutDialog` has a matching handler.

Please review the code.

Thank you!
